### ISSUE TITLE
Google Analytics integration

### ISSUE THREAD

**Issue by jacopoch** [Simbul/baker#332]
Hi, 
I think this has been addressed by someone already, but how exactly you implement Google Analytics in Baker?

Cheers,

Jacopo

**Comment by jacopoch** [comment]
I explain better: I would just like to know when the application gets launched, if it makes this any easier :)
I know how to do it in a normal iOS app, but as I am no xCode developer I fail in integrating it into my baker based magazine..

**Comment by ghost** [comment]
Have you looked at the appdelegate? Also - you might think about putting in the localytics library - it's a little richer than google analytics. You can create your own triggers and events.

**Comment by jacopoch** [comment]
Ok, but what about tracking the single pages with localytics?
I've checked it out vs google analytics, and I'll go for it, it has more features, thanks for the tip ;)

**Comment by ghost** [comment]
If I were to guess, the authors of this library are getting defining moments every time they read this stuff!

If I were in charge of the API, I would consider the following:

BakerFramework is a wonderful bridge between web pages and iOS. I like it a great deal.

The snapshot, updating features, getting script-filled html pages to work in iOS in an epub-like format is an awesome feat - so the bridge works great!

I think you can get the actual html page somewhere in the curr page code (somewhere around the snapshot stuff.)

Going forward, it would be helpful to expose those variables a bit in documentation. It appears that people are getting really sophisticated!

**Comment by jacopoch** [comment]
I think that since the framework is so flexible is quite obvious people want to do more and more with it.

We are almost ready to press the first issue of a monthly magazine of about 100 pages with Baker for example :)

**Comment by folletto** [comment]
I know it might sound rough, and maybe it doesn't work, but it might be possible to just add the Google Analytics code in every page of the book... And maybe trigger a custom javascript event to track the single page. 

Did anyone test it yet? :D
It might be an awesome think if it works... :)

**Comment by jacopoch** [comment]
It doesn't work unfortunately, it is not allowed to execute JavaScript like that from the app directly.
But I'll give a try with localytics tonight, I can live with just the open the application tracking for the moment: I really need to press the issue before 15th of march :)

**Comment by jacopoch** [comment]
Jeep, no analytics love yet.

If somebody has integrated already please share the valuable piece of information so that we can document it :)

**Comment by ghost** [comment]
I struggled with actually stating this in a reply but here goes:

The challenge with OpenSource growth is in its developers inability to
document their own code.

Lots of developers want to reuse, extend, modify OpenSource. Whether or not
projects are extended by its audience is predicated by two things: Whether
they can confidently modify existing code - within a construct and have the
"audacity" to share it and whether or not they have the skills.

The skills part is frequently driven by their ability to understand what is
in front of them.  Sure, they might understand what is being done line by
line; or they can scan well-commented code and go to the right place.
 Whether they do the latter is driven by their ability to devote and
willingness to invest the time.

The last sentence is me.  Nonetheless, I'm looking for the right place to
give you a good example. I've used Localytics and it works wonderfully.

**Comment by folletto** [comment]
_"The challenge with OpenSource growth is in its developers inability to document their own code."_

As you might guess, I disagree, but you say it better later:

_"Whether they do the latter is driven by their ability to devote and willingness to invest the time."_

It's not willingness, otherwise this project wouldn't be here at all. It's a matter of time. Our time is limited, as it is with any opensource project, where "time" is the primary and absolute constraint. So, it's very often a choice between writing code (and fixing, adding features or improving) and documentation, community, pr and so on.

I personally think that we did a quite amazing work in Baker, given the time we had.

We can improve? Of course. But since our donations flow doesn't allow us to dedicate more time, the only extra time here could come from the community.

This means... that's up to you.

If you want, you can easily create a page on the wiki above with all the details you think are necessary, polish and expand it. Opensource isn't just about "the developers", but it's about the **community**. If the community isn't willing to support the whole project - and that doesn't mean just money - with documentation, contributions, feedback, and so on, the developers are in a really tight spot and they have to make difficult choices and compromises.

So, to rephrase myself:

I personally think that we did a quite amazing work in Baker, given the time we had and that's at least 50% thanks to the community here. :)

**Comment by ghost** [comment]
I think if you add the load library stuff in the app delegate as in the
Localytics docs and then look at the RootViewController - around Line 542
#pragma mark "Loading"

Down at line 568, it's loading a page (in a rect view).

currentPageNumber is an integer, you can get that - otherwise, I believe
pages objectAtIndex is the actual localURL (NSURL datatype)..

Short of any documentation and loading up the Localytics library, I think
that's your "insertion" point.

Hope that helps.

**Comment by ghost** [comment]
Hi, without getting into a huge debate....

(Pardon my directness, it avoids people guessing what's on my mind...LOL)

I agree that you did INCREDIBLE WORK!!!

After being in code since 1999, I find that if I don't document everything
as I go, I find that I have to retrain myself each and every time I go back
into the code.  The time that takes = the time it takes to document as you
go.

The challenge with any new project is that you cycle through adding code
and modifying what you have over and over again until it works. That really
gets in the way!!!

At any rate, we have what we have and I am truly appreciative of what you
have done.  I hope in some small way I can continue to contribute to the
community.

Best Wishes, jws

**Comment by jacopoch** [comment]
That's why there is the community,
we can help commenting the procedures / code and creating documentation :)

My aim was to contribute to create some sort of description here so that it's easily accessible and others willing to use it could receive help.

If we manage we'll deliver another improvement to this great open source project, if we won't, let's not start an argument on open source!

J

**Comment by folletto** [comment]
Thanks. :)

**Comment by ghost** [comment]
Yep. Agree!

David, what would have saved me the most time is this:

"Given a JSON file with HTML, count page numbers, pull in content into a
CGRect, along the way zipping screen shots to increase page view
performance.

The major portions of code you should know are:
Root View Controller does the heavy lifting for pages.
Properties does retrieval for pages and yada yada.
Interceptor handles the popup view below."

And I'm still not sure that's right! LOL

**Comment by folletto** [comment]
Check issue #59 and #12 (guess also how old they are from their number) and imagine the decision we took instead of doing these two things. ;)

However, I already started commenting the code every time I touch some function. It just... takes time. :)

**Comment by tstegart** [comment]
I've been using Flurry for analytics and I've just got it working with page views by placing it inside the root view controller.m code. Its still got a few problems, most noticeably that it doesn't count a page view when you open the app, only when you move through the pages after opening. Also, I can't get it to tell me which page was viewed, only that a page was viewed. I've also got it to track outbound links (our magazine is working on getting advertising and need to track clicks) Again, I can track outbound links but not which ones yet. Flurry does have a way of passing information out regarding these things, i just don't know how to implement it. (By the way, the rest of Flurry works fine, I can track users, sessions, countries, etc)

The Flurry interface is really great and I'm excited to try work on it more. If I get it to work I will try update the wiki. I know Google analytics also has an iOS SDK that works in a similar way. Maybe we can all try figuring this out as i know it will be important to magazines that wish to use advertising. I would suggest limiting ourselves right now to Flurry and Google, I know they're both popular. (By the way, to avoid confusion, Google has TWO analytics implementations, the web code and the iOS code. The web code does not work inside the app, so you can't just place it in your HTML files [yes, I tried]. The iOS code is placed inside Xcode files and works the same as Flurry. Please be exact so not to confuse people)   

If anyone at Baker is interested in this and willing to work with me I would be happy to explain what i know, what the problems are and how they might be fixed. You guys know the code and I'm sure it would go quickly. Maybe we can place comments in the code that will allow people to add it on their own without actually building it into Baker. 

P.S. - This is where I placed the Flurry code for web views in rootviewcontroller.m. You also need to place it in the main thread to enable Flurry.     

```
- (BOOL)loadWebView:(UIWebView*)webView withPage:(int)page {

NSString *path = [NSString stringWithString:[pages objectAtIndex:page - 1]];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        NSLog(@"• Loading: book/%@", [[NSFileManager defaultManager] displayNameAtPath:path]);

        // Flurry Analytics page view logging
        [FlurryAnalytics logPageView];

        // Flurry Analytics event logging for pages
        [FlurryAnalytics logEvent:@"Page_Read"];

        [webView loadRequest:[NSURLRequest requestWithURL:[NSURL fileURLWithPath:path]]];
        return YES;

    }
    return NO;
}  
```

**Comment by folletto** [comment]
@tstegart One suggestion that I might give you: if while doing this you are able to make it modular, then we have one more chance to incorporate it in our codebase. :)

Btw, thanks to your work I changed this issue from "Support" to "Feature", since what you are doing seems plausible and even if we end up not supporting a specific analytics tool, we might provide callbacks - or building a delegate. ;)

**Comment by tstegart** [comment]
What do you mean modular? 

I will try and get anyone interested in this the background on both Flurry and Google so people can look at the documentation and see what the best approach is. We're launching our app this week so I won't have it for you this week, but maybe next week. I will post a also summary of the issues and an approach to solving them.

**Comment by jacopoch** [comment]
I am going to share my knowledge about Google analytics integration in the next days,
thanks in the meanwhile for your updates!
I am very busy at the moment and I missed to update you, but good progress has been made to integrate fully GA :)

**Comment by folletto** [comment]
@tstegart  By Modular I mean:
1. Make a function that calls the analytics library
2. From all around the code, call **only** that function, never add blobs of code around. Just the function.

In this way, you will have a clean separation: the analytics code inside a function, and a single line in all the callback points. It will also allow us to see where you need callbacks and who knows, we might find out some time to package up a delegate. ;)

**Comment by kilop** [comment]
Hello.

We are looking for a solution to get the time spent per page. Flurry include this feature?

Thanks

**Comment by tstegart** [comment]
It can. You can fire an event when the user gets to as page, and then fire an event when a user leaves, and then calculate the time of between the event. Its not ideal if you have a TON of events to track, but it will definitely work if you only have a few for now.

**Comment by tstegart** [comment]
Ok folks, here's the scoop on analytics. I've collected the information needed to implement Flurry and Google and will try explain it here. I'm afraid I'm not really much of a coder, but maybe we can work through the best way to implement it going forward as I know a lot of professional magazines will need this to work in order to provide a sustainable business modal.

Overview:

Both Google and Flurry provide SDK's for integrating analytics into your code. Google's info page can be found here: http://code.google.com/apis/analytics/docs/mobile/ios.html#howItWorks and Flurry's info page can be found here: http://support.flurry.com/index.php?title=Analytics/GettingStarted/TechnicalQuickStart

Both Google and Flurry and other analytics platforms provide similar functionality. They generally:
- track the number of times the app is opened;
- try and track the number of users and make guesses about your users;
- track the length of time your app is opened;
- allow you to track custom events
- allow you to track page views
- send this information to their servers to be dissected and visualized

The first three things (tracking number of times you app is used, for how long, and how many users you have) are accomplished automatically when you include the Flurry or Google code in your app. This is done by inserting the code immediately after applicationDidFinishLaunching for both Flurry and Google and following the other simple directions in the Readme file. Flurry and Google collect the information and automatically send it back to be analyzed. This is the the first place Baker needs to accommodate analytics code.

The other two common uses of analytics, custom events and page views are more difficult. Both Google and Flurry provide an automatic way to track page views using UIViewController, UINavigationController or UITabBarController.  I don't think Baker uses these and I couldn't get it to work. Someone can correct me if I'm wrong. 

However, you can track page views another way by calling trackPageView in Google and logPageView in Flurry. These needs to be inserted whenever you want to track a page view. I got this working in a previous version of Baker as described in the code a few posts above this one, but maybe there is a better place. The only important thing to remember is it must only track each page once, and must not track a page view when a user is scrolling through Baker quickly. This is the second place Baker needs to have analytics code.

The third thing, custom even tracking, appears the most difficult. However, its is vitally important to commercial magazines that need to track page views of advertisements and outbound links. The problem with the page view code above is that it tracks a page view, not WHICH page was viewed. Both Google and Flurry allow you to custom track events though, so this should be possible in Baker somehow. 

This is how Flurry describes its event tracking capabilities:

[FlurryAnalytics logEvent:@"EVENT_NAME"];

Use logEvent to count the number of times certain events happen during a session of your application. This can be useful for measuring how often users perform various actions, for example. Your application is currently limited to counting occurrences for 300 different event ids (maximum length 255 characters).

[FlurryAnalytics logEvent:@"EVENT_NAME" withParameters:YOUR_NSDictionary];

Use this version of logEvent to count the number of times certain events happen during a session of your application and to pass dynamic parameters to be recorded with that event. Event parameters can be passed in as a NSDictionary object where the key and value objects must be NSString objects. For example, you could record that a user used your search box tool and also dynamically record which search terms the user entered. Your application is currently limited to counting occurrences for 100 different event ids (maximum length 255 characters). Maximum of 10 event parameters per event is supported.

An example NSDictionary to use with this method could be:

NSDictionary *dictionary =
[NSDictionary dictionaryWithObjectsAndKeys:@"your dynamic parameter value",
@"your dynamic parameter name",
nil];

[FlurryAnalytics logEvent:@"EVENT_NAME" timed:YES];

Use this version of logEvent to start timed event.

[FlurryAnalytics logEvent:@"EVENT_NAME" withParameters:YOUR_NSDictionary timed:YES];

Use this version of logEvent to start timed event with event parameters.

[FlurryAnalytics endTimedEvent:@"EVENT_NAME" withParameters:YOUR_NSDictionary];

Use endTimedEvent to end timed event before app exists, otherwise timed events automatically end when app exists. When ending the timed event, a new event parameters NSDictionary object can be used to update event parameters. To keep event parameters the same, pass in nil for the event parameters NSDictionary object. 

---

Presumeably, Google might work the same way. We would need to figure out how to pass to Flurry and Google the URI of the page that is being viewed and also which URI is clicked on for link tracking.

In my opinion, tracking the names of pages viewed, how long they were viewed and outbound links are the most important data to be tracked for advertisement supported magazines. Of these, tracking the names of the pages viewed and outbound links are far more important than the time spent on each page. This should be the third place Baker needs to accommodate analytics code.

I have not gotten this to work yet and I don't know where in Baker to place the code. If anyone has suggestions, please let me know.

In conclusion, I think the analytics roadmap should look as follows:
1. Identify proper places in Baker to add all analytics code. 
2. Add in analytics code in Baker, test on Flurry and Google.
3. Decide how to make code modular, so that analytics code can be removed from versions of Baker that don't use it.
4. Test modular code.

If anyone has questions or comments or would like to further revise how we should approach this, please post a comment. If you have time, energy, and know how to work on this, please also post. If the Baker folks would like to weigh in on if they think this important enough to spend time on themselves, that would be great too. 
Tom

**Comment by ghost** [comment]
Nicely done!!! Were you able to look at Localytics?  It is significantly easier than this... Essentially, include the lib and add your events.  No NSDicts required (which aren't fun). Lastly, if you look at Parse.com - a great mobile back end - you can spin your own events, then point your own graphing products at the data restfully.  You really are invested here I think and I think you did a great job spelling it all out.  Better than perhaps Flurry!!!

**Comment by tstegart** [comment]
Localytics seems to use the same system, with code needed after applicationDidFinishLaunching and it looks like it also needs NSDictionary events. Take a look at their documentation under "events" here:  http://www.localytics.com/docs/iphone-integration/
I'm guessing most of the analytics programs out there work the same way.

**Comment by jacopoch** [comment]
This is great :)
I'm working on a similar code-based tutorial in my scarce free time as the magazine is approaching to be launched.

But this is super-useful knowledge!

Jacopo

**Comment by tstegart** [comment]
Hey Baker guys,
Is there any way to pass the referrer on to an external website linked to from inside the Baker app? This would help tracking links and clicks on the advertiser's end (or internally, if you're linking to your own website)

**Comment by folletto** [comment]
What do you mean exactly? You mean in the modal window inside Baker?

**Comment by tstegart** [comment]
I mean passing the referrer to the web server serving the page. In the Baker wiki it says this: 

"NOTE: the parameter is used in Baker and will not be used in the URL to contact the server. In the example above the server will receive only http://www.external-site.com."

Is it possible for the server to receive the parameter? That way a website can track how many times people visited them from inside a Baker app by using their own internal analytics to track the referrer.

**Comment by folletto** [comment]
Sure: add whatever parameter you want to the URL. :)

_I asked because the "referrer" technically is a parameter included in the HTTP header. We called the query parameter "referrer" in Baker because it has the same objective. However our parameter isn't technically "the" referrer, but a parameter called "referrer". I should probably add this to the wiki. :D_

**Comment by tstegart** [comment]
Hey everyone. I'm having trouble placing my analytics code, and I was wondering if anyone could help. I'm trying to count pageviews by placing my analytics code to count every time an internal (or external) page is loaded. Previously, as I mentioned in a post above, I placed the code after: 

```
- (BOOL)loadWebView:(UIWebView*)webView withPage:(int)page {

NSString *path = [NSString stringWithString:[pages objectAtIndex:page - 1]];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        NSLog(@"• Loading: book/%@", [[NSFileManager defaultManager] displayNameAtPath:path]); 

\\\ I placed my code here..
```

With the new Baker modal view, everything has been moved around and I'm not really sure how to count page views anymore. I can't seem to figure out when the actual page is loaded, both for internal pages, mail links and external pages that open modal view. Any help would be appreciated.

**Comment by Simbul** [comment]
I'm not sure I'm understanding your requirements correctly, but it looks like your use case may be better covered by placing some code inside the `shouldStartLoadWithRequest` method in `RootViewController.m`.
The method is called when a new page is loaded and when links are clicked. In fact, that's where all our link-handling code is.

**Comment by tstegart** [comment]
Thanks! I'm trying it and I'll let you know how it works after I get my analytics data back.

**Comment by MrSpukhaus** [comment]
Please drop the idea of incuding GA into Baker. In some countries people must be notified if GA is used and must get a right to say "no" to it. If these people don't get informed or can optout of GA usage products containing GA may be subject to expediting. I know this sounds crazy but check it out and come here again if I am wrong / got it wrong. Thanks.

**Comment by folletto** [comment]
Thanks for the feedback. When that gets implemented we have to consider that. :)

**Comment by ghost** [comment]
@MrSpukhaus There are two things that come to mind here about your request:

1) Open Source is exactly that. People have the right to embrace/extend/embellish at will.
2) The concept of "author" applies equally to developer and content creator. As such, either author may be in control of the programmatic content or the digital content riding on top of the programmatic content. Authors have/should/may/can have rights to what is the ultimate form of their content.

That being said, it is truly unfortunate where a small disclosure about usage tracking can solve a whole volume of issues. It is also truly unfortunate that content creators (authors) have to keep in mind that they may be creating something of breadth and scope beyond a normal pedestrian's understanding; in spite of the fact that it was or was not the original authors intent to care.

This is going around Reddit these days - which, at a minimum, got me thinking:

I would much rather be ignorant to the rules of my world, than to be ruthlessly bound to them simply by knowing them.
(I don't know the author of this; however, the question is "What is it like being made of code?")

Does this capture the mindset of the author or the reader or the authorities that create such restrictions? In these days of ACTA/SOPA/PIPA, it appears that the latter has more say than the former.

**Comment by anjimi** [comment]
Google Analytics say: "You must indicate to your users, either in the app itself or in your terms of service, that you reserve the right to anonymously track and report a user's activity inside of your app." Seems fair enough.

Has anyone got analytics working with Baker 4.0? I just had a go at implementing GA but with no success.

I downloaded the most recent GA for iOS (v2.0 beta), I set up a GA app profile and followed the GA instructions to the point of including the bit after applicationDidFinishLaunching as mentioned by @tstegart but I'm not getting any results showing up in google analytics after testing the app. 

I think from the GA instructions, you need to set up measuring screen views for it to work. This is way beyond my limited Xcode knowledge.

https://developers.google.com/analytics/devguides/collection/ios/v2/

**Comment by nin9creative** [comment]
https://github.com/nin9creative/baker

On the master-nin9creative branch I integrated Google Analytics with their recent iOS SDK.  I am only logging the basic "Shelf View" in that branch but it has the SDK integrated.  I also have verified that it works and sends the data to be viewed in the Google Analytics website.

You may want to take a look.

**Comment by anjimi** [comment]
Thanks @nin9creative I will take a look. Just logging the basic Shelf View would be a great start! :)

**Comment by yujulio** [comment]
I've managed to track any book the user is viewing by registering an event Baker logs each time the user open a book using a trackEventWithCategory inside BakerBook.m 

``` objectivec
- (id)initWithBookData:(NSDictionary *)bookData
{
    self = [super init];
    if (self && [self loadBookData:bookData]) {
        NSString *baseID = [self.title stringByAppendingFormat:@" %@", [self.url stringSHAEncoded]];
        self.ID = [self sanitizeForPath:baseID];

        NSLog(@"JSON Parsed successfully, book \"%@ - %@\" created", self.ID, self.title);
        [[GAI sharedInstance].defaultTracker trackEventWithCategory:@"Book Load"
                                                         withAction:nil
                                                          withLabel: (self.title)
                                                          withValue: nil];


        return self;
    }

    return nil;
}
```

With this I can get que title of the book I used in the book.json

And with this:

``` objectivec
- (BOOL)loadWebView:(UIWebView*)webView withPage:(int)page {

    NSString *path = [NSString stringWithString:[pages objectAtIndex:page - 1]];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        NSLog(@"• Loading: book/%@", [[NSFileManager defaultManager] displayNameAtPath:path]);
        [webView loadRequest:[NSURLRequest requestWithURL:[NSURL fileURLWithPath:path]]];
        [[GAI sharedInstance].defaultTracker trackEventWithCategory:@"Page Load"
                                                         withAction:nil
                                                          withLabel: path
                                                          withValue: nil];

        return YES;
    }
    return NO;

    }
```

I can get the page the user is viewing and this is what I get in GA:

``` log
/Users/smartapps02/Library/Application Support/iPhone Simulator/6.1/Applications/D406BD92-1C7A-4699-B7DF-CCF7EDA261C6/Library/Caches/Newsstand/B33DE874-A350-430F-B40F-6AA143B37D15/Indice.html
```

The problem is that I'm only getting the path of the book without the name of the book. Is there anyway to associate them both when the app send the data to GA?

**Comment by Simbul** [comment]
You should be able to call `book.title` inside `loadWebView`.

**Comment by anjimi** [comment]
I'm trying to add GA to our app and struggling as it's beyond my very basic Xcode knowledge/skills!

@yujulio would you be able to assist me by telling me the code you use to initialize GA tracker and where it is placed? I have tried adding the GA files and initialize the tracker as per the GA SDK instructions, I then place the code you have copied above into BakerBook.m but I get a warning on the first GA line `[[GAI sharedInstance].defaultTracker trackEventWithCategory:@"Book Load"`: ! Use of undeclared identifier 'GAI'

Also @Simbul not wanting to get ahead of myself here! but where would one call `book.title` inside `loadWebView`?

**Comment by Simbul** [comment]
Well, `book` is a class variable, so it should e possible to call `book.title` anywhere in `BakerViewController.m`. Or am I misunderstanding the question?

**Comment by yujulio** [comment]
The basic tutorial from GA docs is incomplete, you have to set a default
tracker following this guide first:
https://developers.google.com/analytics/devguides/collection/ios/v2/advanced

2013/4/20 Alessandro Morandi notifications@github.com

**Comment by anjimi** [comment]
Thanks for your help. I tried setting the default tracker as in the GA Advanced Configuration link but I am still getting the same error: Use of undeclared identifier 'GAI'

Here's what I've done so far:

Added the GA files and the frameworks to my Xcode project as per the GA basic overview, imported `GAI.h` in AppDelegate.m and added the GA initialization code:
`// Optional: automatically send uncaught exceptions to Google Analytics.
  [GAI sharedInstance].trackUncaughtExceptions = YES;
  // Optional: set Google Analytics dispatch interval to e.g. 20 seconds.
  [GAI sharedInstance].dispatchInterval = 20;
  // Optional: set debug to YES for extra debugging information.
  [GAI sharedInstance].debug = YES;
  // Create tracker instance.
  id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];`
to `- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions`

I now replaced this line 
`id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];`
with these two lines from the Advanced Config 
`id newTracker = [[GAI sharedInstance]trackerWithTrackingId:@"UA-TRACKING-ID");
[GAI sharedInstance].defaultTracker = newTracker;`

(with UA-TRACKING-ID changed to my GA app ID)

I'm not sure if I am entering the initialization in the right place, or if there is something I have missed. I'm a web developer and pretty new to the objective-c side of things, feeling rather lost here! Any help on how to correctly configure this would be very much appreciated!

**Comment by yujulio** [comment]
Yeah, you're good, you can try it looking at GA if there's any active app.

**Comment by anjimi** [comment]
Ok I got it working pretty good now for book and page views :)

I discovered that I need to `#import "GAI.h"` in BakerBook.m and BakerViewController.m not just in AppDelegate.m; this stopped the errors I was getting: Use of undeclared identifier 'GAI'.

Here's what I have working now:

GA initialization code added to `- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions` in AppDelegate.m:

```
    // Optional: automatically send uncaught exceptions to Google Analytics.
    [GAI sharedInstance].trackUncaughtExceptions = YES;
    // Optional: set Google Analytics dispatch interval to e.g. 20 seconds.
    [GAI sharedInstance].dispatchInterval = 20;
    // Optional: set debug to YES for extra debugging information.
    [GAI sharedInstance].debug = YES;
    // Create tracker instance.
    id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];
    // Set the new tracker as the default tracker, globally.
    [GAI sharedInstance].defaultTracker = tracker;
```

I'm now tracking book and page views as Screen Views rather than Events as this is more appropriate for page views; you can also track Screen Views in Real-Time in GA and don't have to wait for data on Events, which is nice :)

Here's my code in BakerBook.m:

```
- (id)initWithBookData:(NSDictionary *)bookData
{
    self = [super init];
    if (self && [self loadBookData:bookData]) {
        NSString *baseID = [self.title stringByAppendingFormat:@" %@", [self.url stringSHAEncoded]];
        self.ID = [self sanitizeForPath:baseID];

        NSLog(@"[BakerBook] 'book.json' parsed successfully. Book '%@' created with id '%@'.", self.title, self.ID);
        [[GAI sharedInstance].defaultTracker sendView:(self.title)];
        return self;
    }

    return nil;
}
```

This tracks book views, by book title, under Screen Views in GA.

And in BakerViewController.m:

```
- (BOOL)loadWebView:(UIWebView*)webView withPage:(int)page {

    NSString *path = [NSString stringWithString:[pages objectAtIndex:page - 1]];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        NSLog(@"[BakerView] Loading: %@", [[NSFileManager defaultManager] displayNameAtPath:path]);
        [webView loadRequest:[NSURLRequest requestWithURL:[NSURL fileURLWithPath:path]]];

        NSString *GAIpage = [book.title stringByAppendingPathComponent:[[path lastPathComponent] stringByDeletingPathExtension]];
        [[GAI sharedInstance].defaultTracker sendView:GAIpage];

        return YES;
    }
    return NO;
}
```

This tracks page views, by book title/page file name, under Screen Views in GA.

I added a couple lines to `- (void)loadModalWebView:(NSURL *)url` also in BakerViewController.m:

```
    NSString *GAIlink = [book.title stringByAppendingPathComponent:url.absoluteString];
    [[GAI sharedInstance].defaultTracker sendView:GAIlink];
```

This tracks modal views, by book title/modal link url, under Screen Views in GA.

Thanks very much for the pointers @yujulio and @Simbul, I would not have got there without them :)

I will be investigating what else can be tracked in GA, but this is a GREAT start!

**Comment by yujulio** [comment]
Thanks to you too @anjimi really useful using screenview instead events! Thanks.

**Comment by folletto** [comment]
Commit aae5f59a2059bf10e302d0fb55e6db123ae42d10 just pushed implements the needed abstraction to support analytic tools, such as Google Analytics.

Now Baker has the following events tracked:
- BakerApplicationStart
- BakerIssueDownload
- BakerIssueOpen
- BakerIssueClose
- BakerIssuePurchase
- BakerIssueArchive
- BakerSubscriptionPurchase
- BakerViewPage
- BakerViewIndexOpen
- BakerViewModalBrowser

The idea is the following: you should now be able to customize just one file, `BakerAnalyticsEvents.m` and have all the tracking you need. I tested it with Google Analytics but should work also on Countly (given the discussion happening on #1037).

There's also a draft on the wiki with the explanation on how to use it:
https://github.com/Simbul/baker/wiki/How-to-setup-google-analytics-or-similar

We tried to consolidate all the changes you might need to a single file. The overall idea is that if someone creates an integration with a tracking tool, the only required thing is to replace the `BakerAnalyticsEvents.m` with an updated one and everything works.

Give it a go and let us know how it works. :)

**Comment by nin9creative** [comment]
Sweet :)  Was just going to email about this :)

Andrew

**Comment by folletto** [comment]
Has anyone tested this addition on Master yet? :)

**Comment by folletto** [comment]
Soooo.... Nobody? :)
I would have liked a double check, but I guess we'll have to release without. I'm closing this.

**Comment by fturner19** [comment]
@folletto - I just checked it using Countly and it works like a charm, does everything it suppose to do. Great work guys!
![screen shot 2013-09-18 at 10 29 10 am](https://f.cloud.github.com/assets/280993/1165590/e5e53f40-206e-11e3-92a8-71a975f4b66e.png)

**Comment by folletto** [comment]
@fturner19 that. is. great. :)
Thank you! :D

**Comment by jhmac** [comment]
I have added GA to my app. What do I need to do about the EULA to say I am using GA to collect info. I have read that you need to put this sentence,"This application uses Google Analytics to anonymously track usage data within the application.". When I went to add it, it looked like I had to provide a whole new EULA not just add to Apple's.
Has anyone added GA to their app and had it accepted? If so what do you do about the EULA?
thanks

**Comment by folletto** [comment]
This issue is already closed. :)
I would suggest you to open a new one in the new issue tracker, here:
https://github.com/BakerFramework/baker/issues
